src/native/eventpipe/ep-buffer-manager.c

```diff
@@ -1373,7 +1371,29 @@ ep_buffer_manager_get_next_event (EventPipeBufferManager *buffer_manager)
 	// buffer at the same time.
 	ep_timestamp_t stop_timestamp = ep_perf_timestamp_get ();
 	buffer_manager_move_next_event_any_thread (buffer_manager, stop_timestamp);
+
+	EP_SPIN_LOCK_ENTER (&buffer_manager->rt_lock, section1)
+		for (dn_list_it_t it = dn_list_begin (buffer_manager->thread_session_state_list); !dn_list_it_end (it); ) {
+			EventPipeThreadSessionState *session_state = *dn_list_it_data_t (it, EventPipeThreadSessionState *);
+
+			it = dn_list_it_next (it);
+			// if a session_state was exhausted during this sequence point, mark it for deletion
+			if (ep_thread_session_state_get_buffer_list (session_state)->head_buffer == NULL) {
+				// We don't hold the thread lock here, so it technically races with a thread getting unregistered. This is okay,
+				// because we will either not have passed the above if statement (there were events still in the buffers) or we
+				// will catch it at the next sequence point.
+				if (ep_rt_volatile_load_uint32_t_without_barrier (ep_thread_get_unregistered_ref (ep_thread_session_state_get_thread (session_state))) > 0) {
+					dn_list_remove (buffer_manager->thread_session_state_list, session_state);
```

Comment:
I believe we would need to delete the session state in a similar way as done in ep_buffer_manager_write_all_buffers_to_file_v4 or the session. It might make sense to try to isolate the cleanup logic into some shared functions that can be called from both places with similar logic.